### PR TITLE
Source level documentation - part 1

### PR DESCRIPTION
**Pull request by simoncozens**

### CODE REVIEW

**Comment by simoncozens** [comment]
Oh right, it's actually running doctests. That's a bit of a challenge when giving user-friendly examples, but probably a good thing to have...

**Comment by simoncozens** [comment]
While I'm only half-way through documenting the library source, I figure I'd better pause here and open it up for review, because if we wait until I've done it may become hard to merge with any code movement that happens in the middle.

File: Doc/source/agl.rst [hunk 1]
```diff
@@ -1,8 +1,6 @@
...
-###
+######################################
+agl: Interface to the Adobe Glyph List
+######################################
 
 .. automodule:: fontTools.agl
-   :inherited-members:
-   :members:
-   :undoc-members:
+   :members: toUnicode
```
Comment (dscorbett) [inline Doc/source/agl.rst:6]: agl.py also documents the members `UV2AGL` and `AGL2UV`.

File: Lib/fontTools/agl.py [hunk 2]
```diff
@@ -5085,7 +5107,7 @@ def _builddicts():
 def toUnicode(glyph, isZapfDingbats=False):
 	"""Convert glyph names to Unicode, such as 'longs_t.oldstyle' --> u'ſt'
```
Comment (dscorbett) [inline Lib/fontTools/agl.py:5108]: `'longs_t.oldstyle'` and `u'ſt'` are Python code so should be in code spans.

File: Lib/fontTools/cffLib/specializer.py [hunk 3]
```diff
@@ -1,6 +1,17 @@
 # -*- coding: utf-8 -*-
 
-"""T2CharString operator specializer and generalizer."""
+"""T2CharString operator specializer and generalizer.
+
+Postscript glyph drawing operations can be expressed in multiple different
```
Comment (dscorbett) [inline Lib/fontTools/cffLib/specializer.py:5]: “Postscript” should be “PostScript”.

File: Lib/fontTools/cu2qu/cu2qu.py [hunk 4]
```diff
@@ -268,7 +277,10 @@ def cubic_approx_spline(cubic, n, tolerance, _2_3=2/3):
 @cython.locals(max_err=cython.double)
 @cython.locals(n=cython.int)
 def curve_to_quadratic(curve, max_err):
-    """Return a quadratic spline approximating this cubic bezier.
+    """Return a quadratic spline approximating this cubic bezier. Takes a curve
+    (a list of four pairs) and an error value in font units. See the following
+    method for how to interpret the result value.
+
     Raise 'ApproxNotFoundError' if no suitable approximation can be found
```
Comment (dscorbett) [inline Lib/fontTools/cu2qu/cu2qu.py:284]: `ApproxNotFoundError` should be formatted as Python code.

File: Lib/fontTools/cu2qu/cu2qu.py [hunk 5]
```diff
@@ -288,6 +300,22 @@ def curve_to_quadratic(curve, max_err):
...
+    ...   [ (75,50), (120,100), (150,75),  (200,60) ]
+    ... ], [1,1] )
+    [[(50.0, 50.0), (75.0, 75.0), (125.0, 91.66666666666666), (175.0, 75.0), (200.0, 50.0)], [(75.0, 50.0), (97.5, 75.0), (135.41666666666666, 82.08333333333333), (175.0, 67.5), (200.0, 60.0)]]
+
+    The returned splines have "implied oncurve points" suitable for use in
+    TrueType ``glif`` outlines - i.e. in the first spline returned above,
+    the first quadratic segment runs from (50,50) to
+    ( (75 + 125)/2 , (120 + 91.666..)/2 ) = (100, 83.333...).
+
     Raise 'ApproxNotFoundError' if no suitable approximation can be found
```
Comment (dscorbett) [inline Lib/fontTools/cu2qu/cu2qu.py:319]: `ApproxNotFoundError` should be formatted as Python code.

File: Doc/source/cu2qu/index.rst [hunk 6]
```diff
@@ -1,16 +1,38 @@
...
 
-.. automodule:: fontTools.cu2qu
+Instead, this method chops the cubic curve into multiple segments before
+converting each cubic segment to a quadratic, in order to ensure that the
+resulting spline fits within the given tolerance.
+
+The basic curve conversion routines are implemented in the
+:mod:`fontTools.cu2qu.cu2qu` module; the :mod:`fontTools.cu2qu.ufo` module
+applies these routines to all of the curves in a UFO file or files; while the
+:mod:`fontTools.cu2qu.cli` module implements the ``fonttools cu2qu`` command
```
Comment (dscorbett) [inline Doc/source/cu2qu/index.rst:20]: Given that this module is hidden from the documentation, is it appropriate to mention it here?

File: Doc/source/designspaceLib/index.rst [hunk 7]
```diff
@@ -1,16 +1,16 @@
-##############
-designspaceLib
-##############
+########################################################
+designspaceLib: Read, write, and edit .designspace files
+########################################################
 
 MutatorMath started out with its own reader and writer for designspaces.
 Since then the use of designspace has broadened and it would be useful
 to have a reader and writer that are independent of a specific system.
```
Comment (dscorbett) [inline Doc/source/designspaceLib/index.rst:7]: This should document what the module actually does. The reader might interpret “would be useful” to mean the developers plan to implement it but haven’t gotten to it yet.

File: Doc/source/feaLib/index.rst [hunk 8]
```diff
@@ -1,17 +1,40 @@
...
-feaLib
-######
+#########################################
+feaLib: Read/write OpenType feature files
+#########################################
 
-.. toctree::
-   :maxdepth: 1
+fontTools' ``feaLib`` allows for the creation and parsing of Adobe
+Font Development Kit for OpenType feature (.fea) files. The syntax
```
Comment (dscorbett) [inline Doc/source/feaLib/index.rst:6]: Should file name extensions be rendered as code? The documentation is inconsistent.

File: Lib/fontTools/feaLib/parser.py [hunk 9]
```diff
@@ -96,16 +111,18 @@ def parse(self):
         return self.doc_
 
     def parse_anchor_(self):
+        """Parses an anchor in any of the four formats given in the feature
+        file specification (2.e.vii)."""
```
Comment (dscorbett) [inline Lib/fontTools/feaLib/parser.py:115]: It would be helpful for these section references to be links.

File: Lib/fontTools/feaLib/parser.py [hunk 10]
```diff
@@ -273,6 +294,8 @@ def split_glyph_range_(self, name, location):
                 location)
 
     def parse_glyphclass_(self, accept_glyphname):
+        """Parses a glyph class, either named or anonymous, or (if
+        ``accept_glyphname==True``) a glyph name."""
```
Comment (dscorbett) [inline Lib/fontTools/feaLib/parser.py:298]: `accept_glyphname==True` is more specific than the implementation requires. The real requirement is `bool(accept_glyphname)`.

File: Lib/fontTools/misc/loggingTools.py [hunk 11]
```diff
@@ -84,45 +84,66 @@ def format(self, record):
...
+		call is made, they will be reset before the new configuration
+		is applied.
+
+	filename
+		Specifies that a ``FileHandler`` be created, using the specified
+		filename, rather than a ``StreamHandler``.
+
+	filemode
+		Specifies the mode to open the file, if filename is specified
+		(if filemode is unspecified, it defaults to 'a').
```
Comment (dscorbett) [inline Lib/fontTools/misc/loggingTools.py:111]: `'a'` is Python code and should be in a code span.

File: Lib/fontTools/misc/loggingTools.py [hunk 12]
```diff
@@ -84,45 +84,66 @@ def format(self, record):
...
+
+	format
+		Use the specified format string for the handler. This argument
+		also accepts a dictionary of format strings keyed by level name,
+		to allow customising the records appearance for specific levels.
+		The special '*' key is for 'any other' level.
```
Comment (dscorbett) [inline Lib/fontTools/misc/loggingTools.py:117]: `'*'` is Python code and should be in a code span.

File: Lib/fontTools/ttLib/tables/S_V_G_.py [hunk 13]
```diff
@@ -16,44 +16,46 @@
 __doc__="""
 Compiles/decompiles version 0 and 1 SVG tables from/to XML.
 
-Version 1 is the first SVG definition, implemented in Mozilla before Aug 2013, now deprecated.
-This module will decompile this correctly, but will compile a version 1 table
-only if you add the secret element "<version1/>" to the SVG element in the TTF file.
+Version 1 is the first SVG definition, implemented in Mozilla before Aug 2013,
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/S_V_G_.py:19]: “Aug” should be “August”.

File: Lib/fontTools/ttLib/tables/S_V_G_.py [hunk 14]
```diff
@@ -16,44 +16,46 @@
...
+now deprecated. This module will decompile this correctly, but will compile a
+version 1 table only if you add the secret element ``<version1/>`` to the SVG
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/S_V_G_.py:21]: If it is documented as part of the API, in what sense is it a secret?

File: Lib/fontTools/ttLib/tables/otTables.py [hunk 15]
```diff
@@ -1291,26 +1291,28 @@ class ExtendMode(IntEnum):
...
-			Lookup[n] and contents
-				SubTable offset list
-					SubTable[0] and contents
-					...
-					SubTable[n] and contents
-	If the offset to a lookup overflowed (SubTableIndex is None)
-		we must promote the *previous*	lookup to an Extension type.
+	The table layout is::
+
+		GPSO/GUSB
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/otTables.py:1296]: This should be “GPOS/GSUB”.

File: Lib/fontTools/ttLib/tables/otTables.py [hunk 16]
```diff
@@ -1291,26 +1291,28 @@ class ExtendMode(IntEnum):
...
+			Script List
+			Feature List
+			LookUpList
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/otTables.py:1299]: This should be “Lookup List”.

File: Lib/fontTools/varLib/cff.py [hunk 17]
```diff
@@ -157,17 +157,19 @@ def get_private(regionFDArrays, fd_index, ri, fd_map):
 
 def merge_PrivateDicts(top_dicts, vsindex_dict, var_model, fd_map):
 	"""
-	I step through the FontDicts in the FDArray of the varfont TopDict.
-	For each varfont FontDict:
-		step through each key in FontDict.Private.
-		For each key, step through each relevant source font Private dict, and
+	I step through the ``FontDicts`` in the ``FDArray`` of the varfont ``TopDict``.
```
Comment (dscorbett) [inline Lib/fontTools/varLib/cff.py:160]: Documenting a function in first person from the function’s point of view is stylistically inconsistent with the rest of the documentation.

File: Lib/fontTools/varLib/cff.py [hunk 18]
```diff
@@ -599,19 +601,24 @@ def getCommands(self):
 	def reorder_blend_args(self, commands, get_delta_func, round_func):
 		"""
 		We first re-order the master coordinate values.
-		For a moveto to lineto, the args are now arranged as:
+		For a moveto to lineto, the args are now arranged as::
```
Comment (dscorbett) [inline Lib/fontTools/varLib/cff.py:604]: `moveto` and `lineto` should be rendered as code.

File: Doc/source/designspaceLib/index.rst [hunk 19]
```diff
@@ -1,16 +1,16 @@
-##############
-designspaceLib
-##############
+########################################################
+designspaceLib: Read, write, and edit .designspace files
```
Comment (dscorbett) [inline Doc/source/designspaceLib/index.rst:2]: Elsewhere in the documentation they are called “designspace files”, without the dot.

File: Doc/source/designspaceLib/index.rst (hunk 7 above)
Comment (simoncozens) [reply Doc/source/designspaceLib/index.rst:7]: Good point, and the MutatorMath reference isn't necessarily relevant any more.

File: Doc/source/feaLib/index.rst (hunk 8 above)
Comment (simoncozens) [reply Doc/source/feaLib/index.rst:6]: I'll say yes, and make a note to fix up the inconsistencies.

File: Doc/source/cu2qu/index.rst (hunk 6 above)
Comment (anthrotype) [reply Doc/source/cu2qu/index.rst:20]: yes, we don't need to advertise `fontTools.cu2qu.cli` module, it's just internal

File: Lib/fontTools/ttLib/tables/S_V_G_.py (hunk 13 above)
Comment (simoncozens) [reply Lib/fontTools/ttLib/tables/S_V_G_.py:19]: Will fix later, but I haven't got that far in systematically going through the classes yet. This was just fixing a RST syntax nit.

File: Lib/fontTools/varLib/cff.py (hunk 17 above)
Comment (simoncozens) [reply Lib/fontTools/varLib/cff.py:160]: Yep, again I haven't got to this systematically but just wanted to fix up an RST formatting warning. Will fix in part 2.

File: Doc/source/designspaceLib/readme.rst [hunk 20]
```diff
@@ -1076,19 +1074,20 @@ UFOs says.
...
 The designspace format has gone through considerable development. 
 
  -  the format was originally written for MutatorMath.
- -  the format is now also used in fontTools.varlib.
+ -  the format is now also used in fontTools.designspaceLib.
  -  not all values are be required by all implementations.
 
-8.1 Varlib vs. MutatorMath
---------------------------
+8.1 designspaceLib vs. MutatorMath
```
Comment (anthrotype) [inline Doc/source/designspaceLib/readme.rst:1086]: it's actually fontTools.varLib who consumes the designspace files. The designspaceLib module mostly only reads/write the raw data. Keep varLib in here

File: Lib/fontTools/ttLib/tables/otTables.py [hunk 21]
```diff
@@ -1291,26 +1291,28 @@ class ExtendMode(IntEnum):
...
-			Lookup[n] and contents
-				SubTable offset list
-					SubTable[0] and contents
-					...
-					SubTable[n] and contents
-	If the offset to a lookup overflowed (SubTableIndex is None)
-		we must promote the *previous*	lookup to an Extension type.
+	The table layout is::
+
+		GPOS/GUSB
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/otTables.py:1296]: “GUSB” should be “GSUB”.

File: Lib/fontTools/ttLib/tables/otTables.py [hunk 22]
```diff
@@ -1291,26 +1291,28 @@ class ExtendMode(IntEnum):
...
+			Script List
+			Feature List
+			LookUp List
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/otTables.py:1299]: Why use camelcase here?

**Comment by anthrotype** [comment]
This is great, really. However I find it hard to review as it's so big. I think it would be split this into separate PRs for each independent parts.

File: Lib/fontTools/ttLib/tables/otBase.py [hunk 23]
```diff
@@ -42,24 +42,37 @@ def decompile(self, data, font):
...
+
+				call converter's ``write`` method for each item in the value.
+
+					- For simple items, the write method adds a string to the
+					  writer's ``self.items`` list.
+
+					- For ``Struct``/``Table``/``Subtable`` items, it add first adds
+					  new writer to the to the writer's ``self.items``, then calls the
+					  item's compile method. This creates a tree of writers, rooted at the
+					  GUSB/GPOS writer, with each writer representing a table, and the
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/otBase.py:58]: “GUSB” should be “GSUB”.

File: Lib/fontTools/ttLib/tables/otBase.py [hunk 24]
```diff
@@ -42,24 +42,37 @@ def decompile(self, data, font):
...
-			call the getAllData method
-				call _doneWriting, which removes duplicates
-				call _gatherTables. This traverses the tables, adding unique occurences to a flat list of tables
-				Traverse the flat list of tables, calling getDataLength on each to update their position
-				Traverse the flat list of tables again, calling getData each get the data in the table, now that
-				pos's and offset are known.
-
-				If a lookup subtable overflows an offset, we have to start all over.
+
+		Call the compile method for the table.
```
Comment (dscorbett) [inline Lib/fontTools/ttLib/tables/otBase.py:46]: `compile` should be rendered as code.

File: Lib/fontTools/colorLib/builder.py [hunk 25]
```diff
@@ -1,3 +1,9 @@
+"""
+colorLib.builder: Build COLR/CPAL tables from scratch
+
+Example badly needed here.
```
Comment (anthrotype) [inline Lib/fontTools/colorLib/builder.py:4]: I would not add examples for COLRv1 just yet, since it's still a draft proposal

**Pull request closed without merging**